pkg/server/utils/capacity_filter.go

```
@@ -0,0 +1,215 @@
+package utils
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/pkg/errors"
+	"github.com/threefoldtech/grid3-go/deployer"
+	proxyTypes "github.com/threefoldtech/grid_proxy_server/pkg/types"
+	"github.com/threefoldtech/tf-grid-cli/pkg/server/types"
+	"github.com/threefoldtech/zos/pkg/gridtypes"
+)
+
+var (
+	status  = "up"
+	trueVal = true
+)
+
+type Reservation struct {
+	SelectedNodeID uint64
+	WorkloadName   string
+	NeededMemory   uint64
+	NeededStorage  uint64
+}
+
+// Assign chosen NodeIds to cluster node. with both way conversions to/from Reservations array.
+func AssignNodesIDsForCluster(client *deployer.TFPluginClient, cluster *types.K8sCluster) error {
+	// all units unified in bytes
+
+	workloads := []*Reservation{}
+
+	if cluster.Master.NodeID == 0 {
+
+		ms := Reservation{
+			WorkloadName:  cluster.Master.Name,
+			NeededMemory:  uint64(cluster.Master.Memory * int(gridtypes.Megabyte)),
+			NeededStorage: uint64(cluster.Master.DiskSize * int(gridtypes.Gigabyte)),
+		}
+		workloads = append(workloads, &ms)
+	}
+
+	for idx := range cluster.Workers {
+		if cluster.Workers[idx].NodeID == 0 {
+			workloads = append(workloads, &Reservation{
+				WorkloadName:  cluster.Workers[idx].Name,
+				NeededMemory:  uint64(cluster.Workers[idx].Memory * int(gridtypes.Megabyte)),
+				NeededStorage: uint64(cluster.Workers[idx].DiskSize * int(gridtypes.Gigabyte)),
+			})
+		}
+	}
+
+	err := getNodes(client, workloads)
+	if err != nil {
+		return err
+	}
+
+	if cluster.Master.NodeID == 0 {
+		for _, workload := range workloads {
+			if workload.WorkloadName == cluster.Master.Name {
+				if workload.SelectedNodeID == 0 {
```

assignNodes should return an error if one of the reservations failed, else we have established that all reservations were successful and SelectedNodeID != 0